Use AbortController polyfill even when global fetch is available

jpaquim opened the pull request:
The existing code seems to rely on the assumption that if `fetch` is defined, then we're in a browser environment where `AbortController` must also be defined. This is not necessarily the case, in our Node application we ran into an error due to a `fetch = require('node-fetch')` instance already being globally set by another library, although the global `AbortController` was not itself polyfilled.

This commit adds a second conditional branch for setting the `FetchHttpClient`'s internal `this.abortControllerType` independently from `this.fetchType`.

This also addresses the bug described in https://github.com/dotnet/aspnetcore/pull/34881 so it may make sense to close that PR if this one is merged.

jpaquim committed:
Use AbortController polyfill even with global fetch

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=34882) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=34882) before we can accept your contribution.<br/><br/>:x: jpaquim [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=34882)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=34882) it.</sub>

BrennanConroy reviewed (commented):
It would be nice if we could test this somehow

jpaquim committed:
Use webpack magic comments to avoid require in browser

jpaquim committed:
Revert "Use webpack magic comments to avoid require in browser"

This reverts commit 4ea06e5347edfaecc6de2aecb55a27680213c7ba.

jpaquim committed:
Move requireFunc definition inside if guards

jpaquim commented on src/SignalR/clients/ts/signalr/src/FetchHttpClient.ts:
```diff
@@ -20,23 +20,25 @@ export class FetchHttpClient extends HttpClient {
-            const requireFunc = typeof __webpack_require__ === "function" ? __non_webpack_require__ : require;
+        // In order to ignore the dynamic require in webpack builds we need to do this magic
+        // @ts-ignore: TS doesn't know about these names
+        const requireFunc = typeof __webpack_require__ === "function" ? __non_webpack_require__ : require;
```
Thanks for the heads up, can you check again?

I also tried using webpack's `/* webpackIgnore: true */` magic comment to tackle this, as you can see in https://github.com/dotnet/aspnetcore/pull/34882/commits/4ea06e5347edfaecc6de2aecb55a27680213c7ba

Unfortunately, webpack started complaining about `tough-cookie`, with the warning:
```
  external "tough-cookie" 42 bytes [built] [code generated]
    ModuleConcatenation bailout: umd externals can't be concatenated
```

So I decided to go with the simpler route of just moving the `requireFunc` definitions inside the if guards.

jpaquim commented:
@BrennanConroy I addressed your review comment, however I'm a bit lost when it comes to testing, do you have any specific idea of how to do it? Would this use case it fit into any of the existing unit test files (`HttpClient.test.ts`?), or would it make sense to create a new file for `FetchHttpClient` tests in particular?

BrennanConroy commented:
Probably would be a separate `FetchHttpClient` test file.

I was thinking we could try to set a fetch polyfill (doesn't need to be real) without an AbortController and see if trying a send fails, then applying the fix and it doesn't fail (or fails with an error from the fetch polyfill).

BrennanConroy commented:
Looks like
```
it("FetchHttpClient works if fetch replaced and AbortController is not", async () => {
    (global.fetch as any) = () => { throw new Error("error from test"); };
    const httpClient = new FetchHttpClient(NullLogger.instance);

    try {
        await httpClient.post("/");
    } catch (e) {
        expect(e).toEqual(new Error("error from test"));
    }
});
```
Should be a good enough test.

jpaquim committed:
Add regression test

jpaquim commented:
Thanks! Sorry, haven't had the time this week to take a look at it, but I now added the file with your suggested regression test, let me know if you need anything else from my side.

BrennanConroy committed:
Update src/SignalR/clients/ts/signalr/tests/FetchHttpClient.test.ts

BrennanConroy commented:
> let me know if you need anything else from my side

Nope, looks great! Thanks for fixing this 😄
